File: core/src/Serialization.cpp
#include "cmaclib/Serialization.h"

#include <utility>

using namespace CmacLib;

Serialization::Serialization() {}

Serialization::~Serialization(){}

void CmacLib::Serialization::SetString(const std::string& str) {
    serialized = str;
}

void CmacLib::Serialization::SetString(std::string&& str)
{
    this->serialized = std::move(str);
}

std::string CmacLib::Serialization::GetString()
{
    return this->serialized;
}

/// @brief The string from the resulting serialization process.
/// @return serialized string
const char* CCmacLibISerializationGetString(CCmacLibISerialization* pSerialization) {
    return pSerialization->GetString().c_str();
}

/// @brief Indicates if an operation was successful.
/// @return Success.
cmaclib_bool_t CCmacLibISerializationIsSuccessful(
    CCmacLibISerialization* pSerialization
) {
    return pSerialization->IsSuccessful();
}

/// @brief If not successful, a failure message is provided.
const char* CCmacLibISerializationGetMessage(
    CCmacLibISerialization* pSerialization
) {
    return pSerialization->GetMessage().c_str();
}

/// @brief Destructor.
void CCmacLibISerializationDestroy(CCmacLibISerialization** ppSerialization) {
    if (*ppSerialization) {
        delete *ppSerialization;
        *ppSerialization = nullptr;
    }
}

File: wrappers/julia/src/serialization.jl
#!/usr/bin/julia

export ISerialization
"""
Serialization result data.
"""
struct ISerialization end

export ISerializationGetString
"""
The string from the resulting serialization process.

## Arguments
- `pInst::Ptr{ISerialization}`: The pointer to the serialization instance.

## Returns
serialized string
"""
function ISerializationGetString(pInst::Ptr{ISerialization})::Cstring
    return ccall((:CCmacLibISerializationGetString, cmaclib_core),
        Cstring, (Ptr{ISerialization}, ), pInst)
end

export ISerializationIsSuccessful
"""
Indicates if an operation was successful.

## Arguments
- `pInst::Ptr{ISerialization}`: The pointer to the serialization instance.

## Returns
Success.
"""
function ISerializationIsSuccessful(pInst::Ptr{ISerialization})::Bool
    return ccall((:CCmacLibISerializationIsSuccessful, cmaclib_core),
        Bool, (Ptr{ISerialization}, ), pInst)
end

export ISerializationGetMessage
"""
If not successful, a failure message is provided.

## Arguments
- `pInst::Ptr{ISerialization}`: The pointer to the serialization instance.

## Returns
The failure message.
"""
function ISerializationGetMessage(pInst::Ptr{ISerialization})::Cstring
    return ccall((:CCmacLibISerializationGetMessage, cmaclib_core),
        Cstring, (Ptr{ISerialization}, ), pInst)
end

export ISerializationDestroy
"""
Destructor.

## Arguments
- `pInst::Ptr{ISerialization}`: The pointer to the serialization instance.
"""
function ISerializationDestroy(pInst::Ptr{ISerialization})::Nothing
    ccall((:CCmacLibISerializationDestroy, cmaclib_core),
        Cvoid, (Ptr{ISerialization}, ), pInst)
end